src/ui/SortDropdown/SortDropdown.ts

```diff
@@ -0,0 +1,159 @@
+   */
+  constructor(public element: HTMLElement, public options?: ISortDropdownOptions, bindings?: IComponentBindings) {
+    super(element, SortDropdown.ID, bindings);
+
+    this.options = ComponentOptions.initComponentOptions(element, SortDropdown, options);
+
+    this.bind.oneRootElement(InitializationEvents.afterInitialization, () => this.handleAfterInitialization());
+    this.bind.onQueryState(MODEL_EVENTS.CHANGE_ONE, QUERY_STATE_ATTRIBUTES.SORT, (args: IAttributeChangedEventArg) =>
+      this.handleQueryStateChanged(args)
+    );
+    this.bind.onRootElement(QueryEvents.querySuccess, (args: IQuerySuccessEventArgs) => this.handleQuerySuccess(args));
+    this.bind.onRootElement(QueryEvents.queryError, (args: IQueryErrorEventArgs) => this.handleQueryError(args));
+  }
+
+  private handleAfterInitialization() {
+    this.buildDropdown();
+  }
+
+  private buildDropdown() {
+    this.sortItems = this.getSortItems();
+    this.dropdown = new Dropdown(
+      () => this.handleChange(),
+      this.getValuesForDropdown(),
+      value => this.getCaptionForValue(value),
+      l('SortBy')
+    );
+    this.element.appendChild(this.dropdown.getElement());
+    this.update();
+  }
+
+  private getSortItems() {
+    const sortItems: ISortDropdownItem[] = [];
+    const sortElements = $$(this.element).findAll('.CoveoSort');
+
+    each(sortElements, sortEl => {
+      const sortInstance = <Sort>Component.get(sortEl, Sort);
+
+      if (sortInstance.options.sortCriteria.length > 1) {
+        this.logger.warn(
+          `Each Sort component inside a SortDropdown should have only one sort criteria. Not using ${
+            sortInstance.options.caption
+          } in the SortDropdown.`
+        );
+      } else {
+        sortItems.push({
+          caption: sortInstance.options.caption,
+          criteria: sortInstance.options.sortCriteria[0]
+        });
+      }
+    });
+
+    return sortItems;
+  }
+
+  private getValuesForDropdown(): string[] {
+    return this.sortItems.map(item => item.criteria.toString());
+  }
+
+  private handleQueryStateChanged(data: IAttributeChangedEventArg) {
+    this.update();
+  }
+
+  private update() {
+    if (!this.dropdown) {
+      return;
+    }
+
+    const sortCriteria = <string>this.queryStateModel.get(QueryStateModel.attributesEnum.sort);
+    const itemIndex = this.getSortItemIndex(sortCriteria);
+    itemIndex > -1 && this.dropdown.select(itemIndex, false);
+
+    $$(this.dropdown.getElement()).toggleClass('coveo-selected', itemIndex > -1);
+  }
+
+  private handleQuerySuccess(data: IQuerySuccessEventArgs) {
+    data.results.results.length == 0 ? this.hideElement() : this.showElement();
+  }
+
+  private handleQueryError(data: IQueryErrorEventArgs) {
+    this.hideElement();
+  }
+
+  private handleChange() {
+    const selectedValue = this.dropdown.getValue();
+    $$(this.element)
+      .find(`.CoveoSort[data-sort-criteria="${selectedValue}"]`)
+      .click();
```

Comment:
Instead of triggering a click here, the original `Sort` component could be added in `ISortDropdownItem` and `Sort`'s `select` function could be called. Alternatively, a callback which calls `Sort`'s `select` could be added in `ISortDropdownItem` instead.

Reply:
☝️ Yep, we should be calling the instance method on the sort component to perform the job.

There's also the `match()` method on the sort component that you could leverage to find which sort to select().

Reply:
Well, it turns out that a click is needed as `Sort.select()` does not really trigger a query. But, as @btaillon pointed out, I'm going to keep references to the `Sort` instances and click the related elements directly. 

And there's no need anymore of the `ISortDropdownItem` interface, keeping reference to the components will be enough.

Reply:
No, we should really not be doing those kind of "click" simulation. 

Either the code in the sort dropdown trigger the whole selection loop itself: 

```
In: SortDropdown.ts : 

private handleChange () { 
   [ ... ]
   sort.select();
   logUA();
   executeQuery()
   [ ... ]
}
```

or, we add a new public method on the sort component that exposes the full loop.

```
In: Sort.ts

public selectAndExecuteQuery() {

   this.select()
   logUA();
   executeQuery();
}


private handleClick() {
    this.selectAndExecuteQuery();
}


In: SortDropdown.ts 

private handleChange () { 
   [ ... ]
   sort.selectAndExecuteQuery();
   [ ... ]
} 
```
I would vote for option # 2.

Reply:
Nice, makes sense!